### PR TITLE
pfunit: fix max_array_rank for 4+, line len

### PR DESCRIPTION
**Pull request by bd4**

### CODE REVIEW

**Comment by bd4** [comment]
Reference for MAX_RANK being removed in 4.x:
https://github.com/Goddard-Fortran-Ecosystem/pFUnit/tree/3.3.3#configuration

Some old readme's seem to suggest new name is max_array_rank, but can confirm it is MAX_ASSERT_RANK here:
https://github.com/Goddard-Fortran-Ecosystem/pFUnit/blob/v4.0.0/CMakeLists.txt

**Comment by mathomp4** [comment]
@bd4 I now realize I might have screwed up the `package.py` for pFUnit when I tried to update it. Well obviously I broke pFUnit 3 as you found out. But, pFUnit 4 gets its CMake flags for, say, GNU from [`cmake/GNU.cmake`](https://github.com/Goddard-Fortran-Ecosystem/pFUnit/blob/main/cmake/GNU.cmake) (which has the flag you added in question). Thus, I think any lines in the package should apply only to pFUnit 3. 

I have no real way to test pFUnit 3, but could you test changes a la:

https://github.com/mathomp4/spack/compare/develop...mathomp4:spack:fix-up-pfunit

That is:
```diff
diff --git a/var/spack/repos/builtin/packages/pfunit/package.py b/var/spack/repos/builtin/packages/pfunit/package.py
index 44595bafdb..c167a33300 100644
--- a/var/spack/repos/builtin/packages/pfunit/package.py
+++ b/var/spack/repos/builtin/packages/pfunit/package.py
@@ -113,20 +113,25 @@ def cmake_args(self):
             self.define_from_variant("BUILD_SHARED", "shared"),
             "-DCMAKE_Fortran_MODULE_DIRECTORY=%s" % spec.prefix.include,
             self.define_from_variant("BUILD_DOCS", "docs"),
-            "-DMAX_ASSERT_RANK=%s" % spec.variants["max_array_rank"].value,
         ]

-        if self.spec.satisfies("%gcc@10:"):
-            args.append("-DCMAKE_Fortran_FLAGS_DEBUG=-g -O2 -fallow-argument-mismatch")
-
         if spec.satisfies("@4.0.0:"):
             args.append("-DSKIP_MPI=%s" % ("YES" if "~mpi" in spec else "NO"))
             args.append("-DSKIP_OPENMP=%s" % ("YES" if "~openmp" in spec else "NO"))
             args.append("-DSKIP_FHAMCREST=%s" % ("YES" if "~fhamcrest" in spec else "NO"))
             args.append("-DSKIP_ESMF=%s" % ("YES" if "~esmf" in spec else "NO"))
+            args.append("-DMAX_ASSERT_RANK=%s" % spec.variants["max_array_rank"].value)
         else:
+            if self.spec.satisfies("%gcc@10:"):
+                args.append("-DCMAKE_Fortran_FLAGS_DEBUG=-g -O2 -fallow-argument-mismatch")
+
+            if self.spec.satisfies("%gcc@5:"):
+                # prevents breakage when max_array_rank is larger than default
+                args.append("-DCMAKE_Fortran_FLAGS=-ffree-line-length-none")
+
             args.append(self.define_from_variant("MPI", "mpi"))
             args.append(self.define_from_variant("OPENMP", "openmp"))
+            args.append("-DMAX_RANK=%s" % spec.variants["max_array_rank"].value)

         if spec.satisfies("+mpi"):
             args.extend(
```

If that works for you, could you move the two "change my flags" lines into the "if not pFUnit 4" block?

**Comment by bd4** [comment]
There was still a 512 line limit until 4.2.2:
https://github.com/Goddard-Fortran-Ecosystem/pFUnit/commit/cc5ac90e8081708ed58ec3c172119514d9464bdb

I re-arranged so the flags are only used when needed (hopefully). Are spec version ranges inclusive or exclusive for the end of the range? (i.e. should it be "@:4.2.2" or "@:4.2.1", if the issues was fixed and the flag is not needed for 4.2.2)

**Comment by mathomp4** [comment]
> There was still a 512 line limit until 4.2.2: [Goddard-Fortran-Ecosystem/pFUnit@cc5ac90](https://github.com/Goddard-Fortran-Ecosystem/pFUnit/commit/cc5ac90e8081708ed58ec3c172119514d9464bdb)
> 
> I re-arranged so the flags are only used when needed (hopefully). Are spec version ranges inclusive or exclusive for the end of the range? (i.e. should it be "@:4.2.2" or "@:4.2.1", if the issues was fixed and the flag is not needed for 4.2.2)

Oohh. That's a good spot!

And as for your question, I think version ranges are [*inclusive*](https://spack.readthedocs.io/en/latest/packaging_guide.html#version-ranges), so you'd want `@:4.2.1` as that would be "up to and including 4.2.1".

I think. I'm still new at this. Perhaps we should summon @tldahlgren (who helped shepherd through #34476 for me).

**Comment by bd4** [comment]
You are right, is inclusive:
https://spack.readthedocs.io/en/latest/packaging_guide.html#version-ranges

Pushed change.

**Comment by mathomp4** [comment]
> You are right, is inclusive: https://spack.readthedocs.io/en/latest/packaging_guide.html#version-ranges
> 
> Pushed change.

Thanks. I'm currently trying to install spack on my home laptop. I did have an *old* version of spack but I don't trust it, so I'm starting from scratch. Unfortunately, that means building gcc, etc. so I can build pfunit. So...yeah. Hopefully I can test this "soon", where soon depends on my old Macbook's processor. Gotta love building gcc from scratch! 😄

**Comment by bd4** [comment]
> > You are right, is inclusive: https://spack.readthedocs.io/en/latest/packaging_guide.html#version-ranges
> > Pushed change.
> 
> Thanks. I'm currently trying to install spack on my home laptop. I did have an _old_ version of spack but I don't trust it, so I'm starting from scratch. Unfortunately, that means building gcc, etc. so I can build pfunit. So...yeah. Hopefully I can test this "soon", where soon depends on my old Macbook's processor. Gotta love building gcc from scratch! smile

Yea amazing how long it takes even on modern processors - one way to understand how large and complex some of the software we use is. I remember when I used gentoo linux for a while and compiled everything, ain't nobody got time for that now. For my purposes, I have just been using a modified pfunit.py in my spack checkout, so not urgent, just hoping to get changes included in next release. We actually migrated to pFUnit 4 for GENE, so aren't reliant on 3.x, but nice to fix for when I want to build/compare an old version.

**Review by mathomp4 — APPROVED** [review]
Well, I'm currently having an issue where I can't install `gcc` with spack (see #34728). Until that's solved I'm a bit out of sorts. 

But for now this looks good to me and more importantly works for *you*. I'll have opportunities soon--I hope--to test.

File: var/spack/repos/builtin/packages/pfunit/package.py [hunk 1]
```diff
@@ -113,20 +113,26 @@ def cmake_args(self):
...
-
         if spec.satisfies("@4.0.0:"):
             args.append("-DSKIP_MPI=%s" % ("YES" if "~mpi" in spec else "NO"))
             args.append("-DSKIP_OPENMP=%s" % ("YES" if "~openmp" in spec else "NO"))
             args.append("-DSKIP_FHAMCREST=%s" % ("YES" if "~fhamcrest" in spec else "NO"))
             args.append("-DSKIP_ESMF=%s" % ("YES" if "~esmf" in spec else "NO"))
+            args.append("-DMAX_ASSERT_RANK=%s" % spec.variants["max_array_rank"].value)
         else:
+            if spec.satisfies("%gcc@10:"):
+                args.append("-DCMAKE_Fortran_FLAGS_DEBUG=-g -O2 -fallow-argument-mismatch")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/pfunit/package.py:126]: Why is this even setting -O2 in debug?!

**Review by haampie — APPROVED** [review]

File: var/spack/repos/builtin/packages/pfunit/package.py (hunk 1 above)
Comment (bd4) [reply var/spack/repos/builtin/packages/pfunit/package.py:126]: @germasch do you remember why you used O2 here?

**Pull request merged**